File: api/app/schemas/catalog.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TemplateCreate(BaseModel):
    name: str
    description: str | None = None
    category: str | None = None
    chart_repo: str | None = Field(None, alias="chartRepo")
    chart_name: str | None = Field(None, alias="chartName")
    chart_version: str | None = Field(None, alias="chartVersion")
    default_values: dict | None = Field(None, alias="defaultValues")
    values_schema: dict | None = Field(None, alias="valuesSchema")
    model_config = ConfigDict(populate_by_name=True)


class TemplateResponse(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    category: str | None = None
    chart_repo: str | None = Field(None, alias="chartRepo")
    chart_name: str | None = Field(None, alias="chartName")
    chart_version: str | None = Field(None, alias="chartVersion")
    default_values: dict | None = Field(None, alias="defaultValues")
    values_schema: dict | None = Field(None, alias="valuesSchema")
    created_at: datetime = Field(..., alias="createdAt")
    model_config = ConfigDict(from_attributes=True, populate_by_name=True, by_alias=True)


class DeployRequest(BaseModel):
    template_id: UUID = Field(..., alias="templateId")
    app_name: str = Field(..., alias="appName")
    values: dict | None = None
    model_config = ConfigDict(populate_by_name=True)


class DeployResponse(BaseModel):
    message: str
    application_name: str = Field(..., alias="applicationName")
    namespace: str
    template_name: str = Field(..., alias="templateName")
    model_config = ConfigDict(populate_by_name=True, by_alias=True)

File: portal/src/api/catalog.ts
import { get, post, del } from "./client";

export interface CatalogTemplate {
  id: string;
  name: string;
  description: string | null;
  category: string | null;
  chartRepo: string | null;
  chartName: string | null;
  chartVersion: string | null;
  defaultValues: Record<string, unknown> | null;
  valuesSchema: Record<string, unknown> | null;
  createdAt: string;
}

export async function listTemplates(): Promise<CatalogTemplate[]> {
  return get<CatalogTemplate[]>("/api/v1/catalog/templates");
}

export async function getTemplate(id: string): Promise<CatalogTemplate> {
  return get<CatalogTemplate>(`/api/v1/catalog/templates/${id}`);
}

export async function createTemplate(data: Partial<CatalogTemplate>): Promise<CatalogTemplate> {
  return post<CatalogTemplate>("/api/v1/catalog/templates", data);
}

export async function deleteTemplate(id: string): Promise<void> {
  return del(`/api/v1/catalog/templates/${id}`);
}

export async function deployFromTemplate(
  slug: string,
  tier: string,
  data: { templateId: string; appName: string; values?: Record<string, unknown> }
): Promise<{ message: string; applicationName: string; namespace: string; templateName: string }> {
  return post(`/api/v1/teams/${slug}/environments/${tier}/deploy`, data);
}
